app/forms/vpc-router-route-common.tsx

```diff
@@ -0,0 +1,105 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+
+import type { UseFormReturn } from 'react-hook-form'
+
+import type { RouterRouteCreate, RouterRouteUpdate } from '~/api'
+import { DescriptionField } from '~/components/form/fields/DescriptionField'
+import { ListboxField } from '~/components/form/fields/ListboxField'
+import { NameField } from '~/components/form/fields/NameField'
+import { TextField } from '~/components/form/fields/TextField'
+import { Message } from '~/ui/lib/Message'
+
+export type RouteFormValues = RouterRouteCreate | Required<RouterRouteUpdate>
+
+export const routeFormMessage = {
+  vpcSubnetNotModifiable:
+    'Routes of type VPC Subnet within the system router are not modifiable',
+  internetGatewayTargetValue:
+    'For ‘Internet gateway’ targets, the value must be ‘outbound’',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L201-L204
+  noNewRoutesOnSystemRouter: 'User-provided routes cannot be added to a system router',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L300-L304
+  noDeletingRoutesOnSystemRouter: 'System routes cannot be deleted',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L136-L138
+  noDeletingSystemRouters: 'System routers cannot be deleted',
+}
+
+type RouteFormFieldsProps = {
+  form: UseFormReturn<RouteFormValues>
+  isDisabled?: boolean
+}
+export const RouteFormFields = ({ form, isDisabled }: RouteFormFieldsProps) => {
+  const { control } = form
+  const targetType = form.watch('target.type')
+  return (
+    <>
+      {isDisabled && (
+        <Message variant="info" content={routeFormMessage.vpcSubnetNotModifiable} />
+      )}
+      <NameField name="name" control={control} disabled={isDisabled} />
+      <DescriptionField name="description" control={control} disabled={isDisabled} />
+      <ListboxField
+        name="destination.type"
+        label="Destination type"
+        control={control}
+        items={[
+          // VPCs can not be specified as a destination in custom routers
+          // https://github.com/oxidecomputer/omicron/blob/4f27433d1bca57eb02073a4ea1cd14557f70b8c7/nexus/src/app/vpc_router.rs#L363
+          { value: 'ip', label: 'IP' },
+          { value: 'ip_net', label: 'IP network' },
+          { value: 'subnet', label: 'Subnet' },
+        ]}
```

Comment:
Inlining these is clearly more readable, but an advantage of the old way was that we would get a type error if the list of destination or target types changed in the generated API.

```ts
// VPCs can not be specified as a destination in custom routers
// https://github.com/oxidecomputer/omicron/blob/4f27433d1bca57eb02073a4ea1cd14557f70b8c7/nexus/src/app/vpc_router.rs#L363
const destTypes: Record<Exclude<RouteDestination['type'], 'vpc'>, string> = {
  ip: 'IP',
  ip_net: 'IP network',
  subnet: 'Subnet',
}

// Subnets and VPCs cannot be used as a target in custom routers
// https://github.com/oxidecomputer/omicron/blob/4f27433d1bca57eb02073a4ea1cd14557f70b8c7/nexus/src/app/vpc_router.rs#L362-L368
const targetTypes: Record<Exclude<RouteTarget['type'], 'subnet' | 'vpc'>, string> = {
  ip: 'IP',
  instance: 'Instance',
  internet_gateway: 'Internet gateway',
  drop: 'Drop',
}

const toItems = (mapping: Record<string, string>) =>
  Object.entries(mapping).map(([value, label]) => ({ value, label }))
```

Reply:
Yeah, was weighing that. Will revert to get the types back in.